File: Mobile/GoodsTracker/app/src/main/java/com/example/jefferson/goodstracker/Communication/PayLoad.java
package com.example.jefferson.goodstracker.Communication;

/**
 * Created by Jefferson on 08/07/2017.
 */

public class PayLoad {

    public final int    LEN_MAX_PAYLOAD = 256;
    private String      data            = "";

    public PayLoad(){

        clear();
    }

    public void append(char b)
    {
        data += b;
    }

    public void append(String b)
    {
        data += b;
    }

    public void append(boolean b)
    {
        data += String.valueOf((b?1:0));
    }


    public void append(double b) {

        //TODO verificar o que o parametro G faz na conversao string
        //data += b.toString("G");

        data += Double.toString(b);
    }

    public int length()
    {
        return data==null? 0:data.length();
    }

    public boolean isFull()
    {
        return length() >= LEN_MAX_PAYLOAD;
    }

    public boolean isEmpty()
    {
        return length() <= 0;
    }

    public String str() {

        return String.format("%03d",length()) + CONST_COM.CHAR.SEPARATOR + data + CONST_COM.CHAR.SEPARATOR;
    }

    public void clear()
    {
        data = "";
    }

    public char[] toCharArray() {

        return str().toCharArray();
    }

    public String getData() {
        return data;
    }

    public void setData(String data) {
        this.data = data;
    }
}

File: sw/Communication/PayLoad.cs
namespace GoodsTracker
{
    class PayLoad
    {
        public const int    LEN_MAX_PAYLOAD = 256;
        private string      data;

        internal string Data { get => data; set => data = value; }

        internal void Append(char b)
        {
            data += b;
        }

        internal void Append(string b)
        {
            data += b;
        }

        internal void Append(bool b)
        {
            data += (b ? 1 : 0).ToString();
        }

        internal void Append(double b)
        {
            data += b.ToString("G");
        }

        internal int Length()
        {
            return data == null ? 0 : data.Length;
        }

        internal bool IsFull()
        {
            return Length() >= LEN_MAX_PAYLOAD;
        }

        internal bool IsEmpty()
        {
            return Length() <= 0;
        }

        internal string str()
        {
            return Length().ToString("D3") + CONST_CHAR.SEPARATOR + data + CONST_CHAR.SEPARATOR;
        }

        internal void Clear()
        {
            data = "";
        }

        internal void setData(CommunicationFrame frame)
        {
            string value = frame.Data;

            if (value.Length > (HeaderFrame.SIZE + 1))
            {
                data = value.Substring((HeaderFrame.SIZE + 1), value.Length - (HeaderFrame.SIZE + 1));
            }
        }

        internal char[] ToCharArray()
        {
            return str().ToCharArray();
        }
    }
}
